### ISSUE TITLE
Improve failure description for `assertStringContainsString()`

### ISSUE THREAD

**Issue by salsan** [sebastianbergmann/phpunit#5128]
| Q                   | A
| --------------------| ---------------
| PHPUnit version     |9.5.27
| PHP version         | 8.1.10
| Installation Method | Composer

#### Summary

When compare strings (teorically) identically but contain space, result is failed

#### Current behavior

```
1) QueryTest::testGetInfo
Failed asserting that 'SANTAGATI SALVATORE' contains "SANTAGATI SALVATORE".

QueryTest.php:30
phpunit:98   
```

#### How to reproduce

```
<?php

declare(strict_types=1);

use PHPUnit\Framework\TestCase;
use Salsan\Members;

final class QueryTest extends TestCase{
    private $paramters = array(
        'id' => '165714',
        'membershipYear' => '2022'
    );

    public function testInit(): object
    {
        $members = new Members\Query($this->paramters);
        $this->assertIsObject($members);

        return $members;
    }

     /**
     * @depends testInit
     */
    public function testGetInfo($members): void
    {
        $info = $members->getList();

        foreach ($info as $member) {
            $this->assertStringContainsString("SANTAGATI SALVATORE", $member['name']);
            $this->assertIsBool($member['isRookie']);     
        }
    }


}
```

#### Expected behavior
The result to be "true"
echo gettype($member["name"]); // result is **string**
echo ($member["name") // result is **SANTAGATI SALVATORE**

I tried with others string without whitespace and work correctly

**Comment by sebastianbergmann** [comment]
Thank you for your report.

Please provide a **minimal**, **self-contained**, reproducing test case that shows the problem you are reporting.

Without such a minimal, self-contained, reproducing test case I will not be able to investigate this issue.

**Comment by salsan** [comment]
I discovered the issue, I try to compare two "**identically**"  string but with different encoding.

```
// PHP version 8.1.10

echo $members[165714]["name"]); // result is "SANTAGATI SALVATORE"

echo strcmp("SANTAGATI SALVATORE", $members[165714]["name"]); //result -1

echo mb_detect_encoding($members[165714]["name"]); // result is "UTF-8"
echo echo mb_detect_encoding("SANTAGATI SALVATORE"); // result is "ASCII"

var_dump(bin2hex($members[165714]["name"])); //  result is string(40) "53414e544147415449c2a053414c5641544f5245"
var_dump(bin2hex("SANTAGATI SALVATORE")); // string(38) "53414e5441474154492053414c5641544f5245"

```

I tried to create dummy data for test case, but all times with **mb_convert_encoding** string stay "ASCII" and not convert to "UTF-8"

**Comment by hopeseekr** [comment]
I don't think this is a bona fide PHPUnit bug.

I think that your UTF-8 string contains a so-called "thin space" character.

You need to output the **bytecode** of the two strings you are comparing. I'm betting that the space in the UTF-8 is *not* ASCII 32, but the 2-byte UTF-8 "Thin Space" 20 09.

See https://www.compart.com/en/unicode/U+2009.

Here's a snippet that demonstrates it:
```php
<?php

$helloWorld = [
	'ascii' => 'Hello, World!',
	'utf-8' => 'Hello, World!',
];

$spacePos = strpos($helloWorld['ascii'], ' ');

$outputHex = function (string $string): string {
	$hex = bin2hex($string);
	$hexParts = str_split($hex, 2);
	$hexString = implode(' ', $hexParts);
	$tab = strlen($string) < 4 ? "\t\t" : "\t";

	return sprintf("|%s|$tab%s", $string, $hexString);
};

foreach ($helloWorld as $encoding => $hello) {

	echo strtoupper($encoding) . ": \"$hello\"\n";

	$parts = [
        $outputHex(mb_substr($hello, 0, $spacePos)),
        $outputHex(mb_substr($hello, $spacePos, 1)),
        $outputHex(mb_substr($hello, $spacePos + 1)),
	];
	print_r($parts);
}
```
Output:
```
ASCII: "Hello, World!"
Array
(
    [0] => |Hello,|	48 65 6c 6c 6f 2c
    [1] => | |		20
    [2] => |World!|	57 6f 72 6c 64 21
)
UTF-8: "Hello, World!"
Array
(
    [0] => |Hello,|	48 65 6c 6c 6f 2c
    [1] => | |		e2 80 89
    [2] => |World!|	57 6f 72 6c 64 21
)
```

**Comment by salsan** [comment]
In this case is No-Break Space (NBSP)  c2 a0

[https://www.compart.com/en/unicode/U+00A0](https://www.compart.com/en/unicode/U+00A0)

I am not an expert, but I believe that PHPUnit works correctly because the strings are different. The issue is with the output is not, in my opinion, user-friendly.

When read

> Failed asserting that 'SANTAGATI SALVATORE' contains "SANTAGATI SALVATORE".

The output is the same, so why did it fail?

I don't know if this possibility exists when using a string. If it fails, the length of the string also appears, as shown in the example below

> Failed asserting that 'SANTAGATI SALVATORE' (40) contains "SANTAGATI SALVATORE" (38).

**Comment by two-plus-two-equals-five** [comment]
Hey all 👋.

I'm trying to determine what the "fix" is here.

Here are some options I've come up with:
1. Make assertStringContainsString() throw if $haystack's encoding !== the $needle's encoding and have its assertion message say as such
(e.g. `Failed asserting that 'SANTAGATI SALVATORE' contains "SANTAGATI SALVATORE" [ascii] encoding does not match [utf-8].`).
2. Keep the assertStringContainsString() as is, but...
A. the exception message contains the encoding at the end of each given string
(e.g. `Failed asserting that 'SANTAGATI SALVATORE' [utf-8] contains "SANTAGATI SALVATORE" [ascii].`).
B. warn specifically about this "thin space"/"No-Break Space" edge case (that @hopeseekr / @salsan pointed out) if the encoding is different and this case is detected.

My thoughts on these options ^:

1. I imagine this route would break a lot of existing tests in many projects that use assertStringContainsString() since encoding is not always the first thing devs might consider when passing a string to the method.
2. ...
A. This seems like the cleanest solution, as the difference in the encoding will suggest to the dev that it's an encoding issue.
B. I'm betting there are more edge cases than this. It might be worth documenting them somewhere.


What do we think? If I've misunderstood, please say!

I'm wondering if other assert*() methods would be affected by this encoding issue... 🤔

**Comment by sebastianbergmann** [comment]
#5505 has been merged for PHPUnit 10.4.
